### DISCUSSION TITLE
Keyboard closing on air tap

### DISCUSSION THREAD

**Discussion by made3** [microsoft/MixedRealityToolkit-Unity#9639] [Q&A] [answered]
Hello everyone!

I am using the TMP Input Field with the TMP_KeyboardInputField on the HoloLens 2. I have also put a pressable HoloLens2 button on top of the input field and as OnClick Interaction event I added the Select function of the TMP Input Field.
It all works fine now, I can open the virtual keyboard with either an air tap or a near interaction with the virtual hand. I can also type etc.
Problems I have now are:

1) Not that big of a problem, but when I open the keyboard via near interaction and close it by pressing the "x" on the keyboard, the input field is still selected. So if I try to perform another near interaction it will not open the keyboard again because it is already selected.

2) The bigger problem is that, once the keyboard is open, it closes when an air tap is performed. This is a problem because usually (I did it myself without realizing and wondered why the keyboard closed) the user relaxes their fingers after opening the keyboard which results into the index finger and thumb going together and ultimatively performing an unwanted air tap which eventually closes the keyboard.
Is there any way to disable the air tap gesture while the keyboard is open? I tried swapping the configuration profiles with a different input profile but it did not result in what I wanted.

Greetings!

**Comment by MaxWang-MS** [comment, marked as answer]
Could you let us know the versions of Unity and MRTK? Both problems could be on the Unity side issue. For the second issue we provide an implementation that exposes a `DisableUIInteractionWhenTyping` property you can use to change the behavior. The implementation is in the `MixedRealityKeyboardBase` class and you can try it by looking at the left hand side of the `HandInteractionExamples` scene and find the keyboard interaction section. Go to the `PressableButtonPlated (1)` GameObject under `KeyboardExample` and scroll down in the inspector to change the value of that setting to see desired outcome.

**Reply by made3** [reply]
Thanks for the quick reply!

Unity version is 2019.3.15f1 and MRTK _**was**_ 2.5.4. I now upgraded to 2.6.1 to have access to the `DisableUIInteractionWhenTyping` property. If this is set to true an air tap on the spatial map is still recognized, right? There is barely any UI the user could interact with, but it is the air taps onto basically nothing which make the keyboard disappear.

I just started the MRTK app on the HoloLens and used the HandInteractionExample scene. They keyboard there does not disappear unless the x button is pressed. This is the behaviour I want, but I don't think it uses the `DisableUIInteractionWhenTyping` property? 
Guess I will just recreate the keyboard etc from HandInteractionExample but with my own buttons etc. I did not want to do that because it seemed easier to just use the TMP_InputField.

**Reply by MaxWang-MS** [reply]
I cannot remember whether that property only disables interactions with Unity UI or other GameObjects in the scene but it should be obvious when you deploy a sample onto the device. And just as you mentioned I do not think we automatically hide the keyboard when user interacts with something else (and this is independent of the property - the property disallows interaction with some/all other GameObjects). The behavior of TMP Input Field is defined by Unity so in this case you could spend some time on it to see if you can customize the behavior programmatically, submit a feature request to Unity or implement your own. Also note there are some known keyboard bugs in some versions of Unity and I would recommend taking a look at #9056 if you ever run into such problems.

**Reply by made3** [reply]
Well, sadly when I only build, deploy and run that one HandInteractions example scene it does not work. The application just closes itself after loading.
I finally also realized that the MRTK does not seem to be causing the problem, its the TMP_InputField. Though, if I would have an option to simply disable/enable distant interactions (air tap) it would be an easy workaround/fix for me.


Edit: I just implemented the exact same thing as in the keyboard example, but my backspace is not working. Seems like he had the same problem: https://github.com/microsoft/MixedRealityToolkit-Unity/issues/8916
But I am on MRTK 2.6.1, so I don't understand the problem. I created a new scene in a different Unity project and tested it but it did not work. Any ideas? This whole keyboard topic is so entangled, one time the problem is in Unity, the other time it is in the MRTK, sometimes it should be fixed already, sometimes not.

**Reply by made3** [reply]
I just created a completely new Unity 2019.3.15f1 project, added the MRTK 2.6.1, built and deployed the HandInteractionExamples scene and it won't start on the HoloLens. It shows the Unity splash screen, then the four loading spheres and then it crashes.

**Reply by MaxWang-MS** [reply]
The backspace issue is a known bug on the Unity side. If you take a closer look at #9056 which I linked above, I mentioned that you need to use a Unity version later than 2019.4.9 but also no later than 2019.4.18 to achieve the result shown in the table on that page. Also note we recommend Unity 2019 LTS (2019.4.x) even if you are not using the keyboard. We definitely want to bring a great keyboard experience but Unity-side bugs are not something we can control.

**Reply by made3** [reply]
Oh god alright. Guess I will try to upgrade my project to 2019.4.18 then... Thank you! I understand that it is not your problem, I am just frustrated that something as simple as keyboard input is not working.
